pegasus-common/src/main/java/com/linkedin/common/callback/Callback.java

```diff
@@ -32,4 +35,13 @@
    * @param e the error
    */
   void onError(Throwable e);
+
+  /**
+   * Called with lazy Throwable creation if the asynchronous operation failed with an error
+   * @param throwableSupplier the supplier to create Throwable when needed
+   */
+  default void onError(Supplier<Throwable> throwableSupplier)
+  {
+    onError(throwableSupplier.get());
+  }
```

Comment:
Do we need this change in the interface? Since this is leverage only inside TimeoutCallback, Why don't we resolve the lambda inside the TimeoutException itself while calling onError?

Reply:
Because we want to postpone the Throwable creation until #onError is called. This can save us a lot of time if onError is never called.

Reply:
We can. But the question is wether we want to limit this optimization to TimeoutCallBack or CallBack at all. If you check how many implementation of CallBack, you'll notice that there are a lot of chances of optimization. Of course most of them have limited usage so no need to do anything for now. But should we leave this option open so that when needed, we do not have to come back to change it?